### PR TITLE
gh-89828: Do not relay the __class__ attribute in GenericAlias

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
`list[int].__class__` returned `type`, and `isinstance(list[int], type)`
returned `True`. It caused numerous problems in code that checks
`isinstance(x, type)`.

#89828

### CODE REVIEW

**Review by Fidget-Spinner — APPROVED** [review]
The code LGTM. I'd prefer to wait a bit to hear what the others have to say though.

**Comment by serhiy-storchaka** [comment]
Alternatively, we can stop relaying any dunder name, except `__name__` and `__qualname__`.

**Review by gvanrossum — APPROVED** [review]
LG!

**Comment by JelleZijlstra** [comment]
I agree with this idea but we need to think about compatibility implications. We can definitely make the change only on main.

**Comment by serhiy-storchaka** [comment]
3.11 has not been released yet. I planned to do this change in 3.11, but there were tasks with more priority before the first beta.

There were many changes in GenericAlias related to PEP 646, and the work is not finished yet. I suspect that keeping this behavior not fixed in 3.11 can cause new 3.11 specific bugs. It will be easier if the code of 3.11 and main be synchronized. @pablogsal

**Comment by pablogsal** [comment]
I'm fine with backporting this as this is in the real of what we can consider a bug if nobody had any concerns. But at least another two core Devs must agree that the backport (by approving it).

Please, also tag me in the backport PR if we decide to move forward with it

**Comment by gvanrossum** [comment]
It's unfortunately we waited so long. Not doing this in 3.11 makes it more likely that there will be code relying on the old behavior (no matter how strange). Alas, the old behavior was released with 3.9 and 3.10 (though some bugs related to this weren't fixed in those releases), so I'm not sure it's that important to fix it in 3.11 -- people are going to have to fix their code regardless. (It would seem that most of the time code that works around this will keep working though -- most of the fixes in this PR are removing workarounds, but that's cleanup -- that code will work with or without the workarounds.

So I'll wait for @pablogsal.

**Comment by serhiy-storchaka** [comment]
Workarounds for known issues in the stdlib were backported to 3.10 and 3.9 (for these for which I had found reproducers). The third-party code can still be affected, of course.

File: Misc/NEWS.d/next/Core and Builtins/2022-06-12-19-31-56.gh-issue-89828.bq02M7.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+:class:`types.GenericAlias` no longer relay the ``__class__`` attribute.
```
Comment (JelleZijlstra) [inline Misc/NEWS.d/next/Core and Builtins/2022-06-12-19-31-56.gh-issue-89828.bq02M7.rst]:
```suggestion
:class:`types.GenericAlias` no longer relays the ``__class__`` attribute.
For example, ``isinstance(list[int], type)`` no longer returns ``True``.
```

Make it clear what the consequence of the change is.

**Review by JelleZijlstra — APPROVED** [review]
The change looks good to me.

I still think we should change this only in 3.12. The old behavior was bad, but it was already that way in 3.9 and 3.10, so there's little harm in waiting a bit longer. However, if others think it's worth changing this in 3.11, I'd be OK with backporting.

**Comment by Fidget-Spinner** [comment]
I think we should backport to 3.11. Not many users use the betas (compared to the actual releases anyways).

**Comment by AlexWaygood** [comment]
I agree with Serhiy. I think the sooner this is fixed, the better, as it will reduce the possibility that third-party code starts to rely on `isinstance(list[int], type)` returning `True`. Otherwise we have another year of this bug percolating through the Python ecosystem and becoming established behaviour.

It's unfortunate this is happening after betas have been released, but I still think it's worth it going into 3.11.

**Comment by AlexWaygood** [comment]
Backporting to 3.11 will also fix #93847, which is a bug that's new in 3.11.

**Comment by gvanrossum** [comment]
Okay what are we waiting for?

**Comment by serhiy-storchaka** [comment]
I was going to add yet few tests and fixes, backport them down to 3.10, then merge this PR and backport it to 3.11. It would be less work than creating different PRs for 3.10 and 3.11-main.

**Comment by serhiy-storchaka** [comment]
Merged, because my computer crashed, and I lost all local changes.

**Pull request merged**